Source/EmbeddedBoundary/WarpXFaceExtensions.cpp

```diff
@@ -0,0 +1,1098 @@
+#include "WarpX.H"
+#include <AMReX_Scan.H>
+#include <AMReX_iMultiFab.H>
+#include <AMReX_MultiFab.H>
+
+/**
+* \brief auxiliary function to count the amount of faces which still need to be extended
+*/
+amrex::Array1D<int, 0, 2>
+WarpX::CountExtFaces() {
+    amrex::Array1D<int, 0, 2> sums{0, 0, 0};
+#ifdef AMREX_USE_EB
+
+    for (int idim = 0; idim < AMREX_SPACEDIM; ++idim) {
+        amrex::ReduceOps<amrex::ReduceOpSum> reduce_ops;
+        amrex::ReduceData<int> reduce_data(reduce_ops);
+        for (amrex::MFIter mfi(*m_flag_ext_face[maxLevel()][idim]); mfi.isValid(); ++mfi) {
+            amrex::Box const &box = mfi.validbox();
+            auto const &flag_ext_face = m_flag_ext_face[maxLevel()][idim]->array(mfi);
+            reduce_ops.eval(box, reduce_data,
+                [=] AMREX_GPU_DEVICE(int i, int j, int k) -> amrex::GpuTuple<int> {
+                    return flag_ext_face(i, j, k);
+                });
+        }
+
+        auto r = reduce_data.value();
+        sums(idim) = amrex::get<0>(r);
+    }
+
+    amrex::ParallelDescriptor::ReduceIntSum(&(sums(0)), AMREX_SPACEDIM);
+#endif
+    return sums;
+}
+
+
+/**
+* \brief This is the main function computing the cell extension. Where possible it computes one-way
+ *       extensions and, when this is not possible, it does eight-ways extensions.
+*/
+void
+WarpX::ComputeFaceExtensions(){
+#ifdef AMREX_USE_EB
+    amrex::Array1D<int, 0, 2> N_ext_faces = CountExtFaces();
+    amrex::ParallelDescriptor::ReduceIntSum(&(N_ext_faces(0)), 3);
+    amrex::Print()<< "Faces to be extended in x:\t" << N_ext_faces(0) <<std::endl;
+    amrex::Print()<< "Faces to be extended in y:\t" << N_ext_faces(1) <<std::endl;
+    amrex::Print()<< "Faces to be extended in z:\t" << N_ext_faces(2) <<std::endl;
+
+    InitBorrowing();
+    amrex::Array1D<int, 0, 2> temp_inds = ComputeOneWayExtensions();
+    amrex::Array1D<int, 0, 2> N_ext_faces_after_one_way = CountExtFaces();
+    amrex::ParallelDescriptor::ReduceIntSum(&(N_ext_faces_after_one_way(0)), 3,
+                                            amrex::ParallelDescriptor::IOProcessorNumber());
+    amrex::Print()<< "Faces to be extended after one way extension in x:\t" <<
+                     N_ext_faces_after_one_way(0) <<std::endl;
+    amrex::Print()<< "Faces to be extended after one way extension in y:\t" <<
+                     N_ext_faces_after_one_way(1) <<std::endl;
+    amrex::Print()<< "Faces to be extended after one way extension in z:\t" <<
+                     N_ext_faces_after_one_way(2) <<std::endl;
+    ComputeEightWaysExtensions(temp_inds);
+    amrex::Array1D<int, 0, 2> N_ext_faces_after_eight_ways = CountExtFaces();
+    amrex::ParallelDescriptor::ReduceIntSum(&(N_ext_faces_after_eight_ways(0)), 3,
```

Comment:
You can probably group the to be reduced arrays of 3 into one and only perform once `ReduceIntSum` over the network.

Reply:
Sorry, I am not sure what you mean by that :)